{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genero\n",
    "\n",
    "Notebook para sacar una subdivicion entre el genero biologico y el eid desde genomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/angie/.local/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "# Import the required libraries for data analysis and visualization\n",
    "import pandas as pd # Para manejar dataframes\n",
    "import numpy as np # Para operaciones matemáticas\n",
    "import os # Para manejar rutas\n",
    "import matplotlib.pyplot as plt # Para graficar\n",
    "import seaborn as sns # Para graficar\n",
    "import re # Para quitar el patron '_0_0'\n",
    "from sklearn.experimental import enable_iterative_imputer  # IterativeImputer\n",
    "from sklearn.impute import IterativeImputer # IterativeImputer\n",
    "from scipy import stats # Para outliers\n",
    "import shap # Para SHAP\n",
    "from sklearn.model_selection import train_test_split # Para dividir el conjunto de datos\n",
    "from sklearn.ensemble import RandomForestClassifier # Para Random Forest\n",
    "from sklearn.metrics import classification_report # Para evaluar\n",
    "from sklearn.preprocessing import LabelEncoder # Para pasar variables categoricas a numericas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>f_22000_0_0</th>\n",
       "      <th>f_22001_0_0</th>\n",
       "      <th>f_22003_0_0</th>\n",
       "      <th>f_22004_0_0</th>\n",
       "      <th>f_22005_0_0</th>\n",
       "      <th>f_22006_0_0</th>\n",
       "      <th>f_22009_0_1</th>\n",
       "      <th>f_22009_0_2</th>\n",
       "      <th>f_22009_0_3</th>\n",
       "      <th>...</th>\n",
       "      <th>f_22024_0_0</th>\n",
       "      <th>f_22025_0_0</th>\n",
       "      <th>f_22026_0_0</th>\n",
       "      <th>f_22027_0_0</th>\n",
       "      <th>f_22028_0_0</th>\n",
       "      <th>f_22029_0_0</th>\n",
       "      <th>f_22030_0_0</th>\n",
       "      <th>f_22007_0_0</th>\n",
       "      <th>f_22008_0_0</th>\n",
       "      <th>f_22182_0_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000012</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.186874</td>\n",
       "      <td>0.188240</td>\n",
       "      <td>0.008535</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-12.4522</td>\n",
       "      <td>2.97502</td>\n",
       "      <td>-1.526670</td>\n",
       "      <td>...</td>\n",
       "      <td>44.7400</td>\n",
       "      <td>99.059</td>\n",
       "      <td>0.97000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0010894A</td>\n",
       "      <td>G03</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000021</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.190092</td>\n",
       "      <td>0.191694</td>\n",
       "      <td>0.001877</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-13.8955</td>\n",
       "      <td>6.01974</td>\n",
       "      <td>-2.011150</td>\n",
       "      <td>...</td>\n",
       "      <td>93.3443</td>\n",
       "      <td>99.477</td>\n",
       "      <td>0.95280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0008357</td>\n",
       "      <td>B08</td>\n",
       "      <td>0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000035</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.190206</td>\n",
       "      <td>0.191802</td>\n",
       "      <td>0.001883</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-11.7626</td>\n",
       "      <td>4.23129</td>\n",
       "      <td>-3.091340</td>\n",
       "      <td>...</td>\n",
       "      <td>30.1373</td>\n",
       "      <td>99.595</td>\n",
       "      <td>0.96018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0008467</td>\n",
       "      <td>C11</td>\n",
       "      <td>1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000044</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.190249</td>\n",
       "      <td>0.191653</td>\n",
       "      <td>0.003119</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-11.2151</td>\n",
       "      <td>4.37638</td>\n",
       "      <td>0.766213</td>\n",
       "      <td>...</td>\n",
       "      <td>29.3400</td>\n",
       "      <td>99.580</td>\n",
       "      <td>0.98000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0011008A</td>\n",
       "      <td>A03</td>\n",
       "      <td>0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000053</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.190025</td>\n",
       "      <td>0.191251</td>\n",
       "      <td>0.002626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-10.4144</td>\n",
       "      <td>2.19996</td>\n",
       "      <td>0.545220</td>\n",
       "      <td>...</td>\n",
       "      <td>22.1800</td>\n",
       "      <td>99.572</td>\n",
       "      <td>0.97000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0013058A</td>\n",
       "      <td>H12</td>\n",
       "      <td>0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid  f_22000_0_0  f_22001_0_0  f_22003_0_0  f_22004_0_0  f_22005_0_0  \\\n",
       "0  1000012         10.0          0.0     0.186874     0.188240     0.008535   \n",
       "1  1000021         -7.0          1.0     0.190092     0.191694     0.001877   \n",
       "2  1000035         -2.0          0.0     0.190206     0.191802     0.001883   \n",
       "3  1000044         57.0          1.0     0.190249     0.191653     0.003119   \n",
       "4  1000053         69.0          0.0     0.190025     0.191251     0.002626   \n",
       "\n",
       "   f_22006_0_0  f_22009_0_1  f_22009_0_2  f_22009_0_3  ...  f_22024_0_0  \\\n",
       "0          1.0     -12.4522      2.97502    -1.526670  ...      44.7400   \n",
       "1          1.0     -13.8955      6.01974    -2.011150  ...      93.3443   \n",
       "2          1.0     -11.7626      4.23129    -3.091340  ...      30.1373   \n",
       "3          1.0     -11.2151      4.37638     0.766213  ...      29.3400   \n",
       "4          1.0     -10.4144      2.19996     0.545220  ...      22.1800   \n",
       "\n",
       "   f_22025_0_0  f_22026_0_0  f_22027_0_0  f_22028_0_0  f_22029_0_0  \\\n",
       "0       99.059      0.97000          NaN          1.0          1.0   \n",
       "1       99.477      0.95280          NaN          1.0          1.0   \n",
       "2       99.595      0.96018          NaN          1.0          1.0   \n",
       "3       99.580      0.98000          NaN          1.0          1.0   \n",
       "4       99.572      0.97000          NaN          1.0          1.0   \n",
       "\n",
       "   f_22030_0_0    f_22007_0_0  f_22008_0_0  \\\n",
       "0          1.0  SMP4_0010894A          G03   \n",
       "1          1.0   SMP4_0008357          B08   \n",
       "2          1.0   SMP4_0008467          C11   \n",
       "3          1.0  SMP4_0011008A          A03   \n",
       "4          1.0  SMP4_0013058A          H12   \n",
       "\n",
       "                                         f_22182_0_0  \n",
       "0  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...  \n",
       "1  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...  \n",
       "2  1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...  \n",
       "3  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,...  \n",
       "4  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set the path to the directory containing the raw data files\n",
    "path = \"../../main/data/UK_BIOBANK_DATA\" \n",
    "\n",
    "#Charge dataframes\n",
    "#Touchscreen\n",
    "genomics_df = pd.read_csv(os.path.join(path, \"genomics_data.tsv\"), sep=\"\\t\")\n",
    "\n",
    "genomics_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Field</th>\n",
       "      <th>FieldID</th>\n",
       "      <th>ValueType</th>\n",
       "      <th>Coding</th>\n",
       "      <th>Category</th>\n",
       "      <th>Participants</th>\n",
       "      <th>Items</th>\n",
       "      <th>Stability</th>\n",
       "      <th>Units</th>\n",
       "      <th>ItemType</th>\n",
       "      <th>Strata</th>\n",
       "      <th>Sexed</th>\n",
       "      <th>Instances</th>\n",
       "      <th>Array</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Link</th>\n",
       "      <th>Path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genotype measurement batch</td>\n",
       "      <td>22000</td>\n",
       "      <td>Categorical single</td>\n",
       "      <td>22000.0</td>\n",
       "      <td>100313</td>\n",
       "      <td>488127</td>\n",
       "      <td>488127</td>\n",
       "      <td>Accruing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Data</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Unisex</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Genotyping was done using two closely related ...</td>\n",
       "      <td>http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...</td>\n",
       "      <td>Genomics &gt; Genotypes &gt; Genotyping process and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genetic sex</td>\n",
       "      <td>22001</td>\n",
       "      <td>Categorical single</td>\n",
       "      <td>9.0</td>\n",
       "      <td>100313</td>\n",
       "      <td>488127</td>\n",
       "      <td>488127</td>\n",
       "      <td>Complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Data</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Unisex</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Sex as determined from genotyping analysis.   ...</td>\n",
       "      <td>http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...</td>\n",
       "      <td>Genomics &gt; Genotypes &gt; Genotyping process and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Heterozygosity</td>\n",
       "      <td>22003</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100313</td>\n",
       "      <td>488127</td>\n",
       "      <td>488127</td>\n",
       "      <td>Accruing</td>\n",
       "      <td>Fraction</td>\n",
       "      <td>Data</td>\n",
       "      <td>Derived</td>\n",
       "      <td>Unisex</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Heterozygosity across a set of high-quality ma...</td>\n",
       "      <td>http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...</td>\n",
       "      <td>Genomics &gt; Genotypes &gt; Genotyping process and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Heterozygosity, PCA corrected</td>\n",
       "      <td>22004</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100313</td>\n",
       "      <td>488127</td>\n",
       "      <td>488127</td>\n",
       "      <td>Ongoing</td>\n",
       "      <td>Fraction</td>\n",
       "      <td>Data</td>\n",
       "      <td>Derived</td>\n",
       "      <td>Unisex</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Heterozygosity after adjusting for ancestry us...</td>\n",
       "      <td>http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...</td>\n",
       "      <td>Genomics &gt; Genotypes &gt; Genotyping process and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Missingness</td>\n",
       "      <td>22005</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100313</td>\n",
       "      <td>488127</td>\n",
       "      <td>488127</td>\n",
       "      <td>Accruing</td>\n",
       "      <td>Fraction</td>\n",
       "      <td>Data</td>\n",
       "      <td>Derived</td>\n",
       "      <td>Unisex</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Missing rate of each sample based on a set of ...</td>\n",
       "      <td>http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...</td>\n",
       "      <td>Genomics &gt; Genotypes &gt; Genotyping process and ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Field  FieldID           ValueType   Coding  \\\n",
       "0     Genotype measurement batch    22000  Categorical single  22000.0   \n",
       "1                    Genetic sex    22001  Categorical single      9.0   \n",
       "2                 Heterozygosity    22003          Continuous      NaN   \n",
       "3  Heterozygosity, PCA corrected    22004          Continuous      NaN   \n",
       "4                    Missingness    22005          Continuous      NaN   \n",
       "\n",
       "   Category  Participants   Items Stability     Units ItemType   Strata  \\\n",
       "0    100313        488127  488127  Accruing       NaN     Data  Primary   \n",
       "1    100313        488127  488127  Complete       NaN     Data  Primary   \n",
       "2    100313        488127  488127  Accruing  Fraction     Data  Derived   \n",
       "3    100313        488127  488127   Ongoing  Fraction     Data  Derived   \n",
       "4    100313        488127  488127  Accruing  Fraction     Data  Derived   \n",
       "\n",
       "    Sexed  Instances  Array  \\\n",
       "0  Unisex          1      1   \n",
       "1  Unisex          1      1   \n",
       "2  Unisex          1      1   \n",
       "3  Unisex          1      1   \n",
       "4  Unisex          1      1   \n",
       "\n",
       "                                               Notes  \\\n",
       "0  Genotyping was done using two closely related ...   \n",
       "1  Sex as determined from genotyping analysis.   ...   \n",
       "2  Heterozygosity across a set of high-quality ma...   \n",
       "3  Heterozygosity after adjusting for ancestry us...   \n",
       "4  Missing rate of each sample based on a set of ...   \n",
       "\n",
       "                                                Link  \\\n",
       "0  http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...   \n",
       "1  http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...   \n",
       "2  http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...   \n",
       "3  http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...   \n",
       "4  http://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=...   \n",
       "\n",
       "                                                Path  \n",
       "0  Genomics > Genotypes > Genotyping process and ...  \n",
       "1  Genomics > Genotypes > Genotyping process and ...  \n",
       "2  Genomics > Genotypes > Genotyping process and ...  \n",
       "3  Genomics > Genotypes > Genotyping process and ...  \n",
       "4  Genomics > Genotypes > Genotyping process and ...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomics_chars_df = pd.read_csv(\n",
    "    os.path.join(path, 'genomics_chars.tsv'),\n",
    "    sep=\"\\t\",\n",
    "    skiprows=5\n",
    ")\n",
    "\n",
    "# Mostrar las primeras filas\n",
    "genomics_chars_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coding_name</th>\n",
       "      <th>code</th>\n",
       "      <th>meaning</th>\n",
       "      <th>concept</th>\n",
       "      <th>display_order</th>\n",
       "      <th>parent_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22000</td>\n",
       "      <td>1000</td>\n",
       "      <td>BiLEVE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22000</td>\n",
       "      <td>-11</td>\n",
       "      <td>UKBiLEVEAX_b11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   coding_name  code         meaning  concept  display_order  parent_code\n",
       "0            9     0          Female      NaN              1          NaN\n",
       "1            9     1            Male      NaN              2          NaN\n",
       "2            1     1             Yes      NaN              1          NaN\n",
       "3        22000  1000          BiLEVE      NaN              1          NaN\n",
       "4        22000   -11  UKBiLEVEAX_b11      NaN              2       1000.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Charge dataframes\n",
    "#Touchscreen\n",
    "genomics_cbkeys_df = pd.read_csv(os.path.join(path, \"genomics_cbkeys.tsv\"), sep=\"\\t\")\n",
    "\n",
    "genomics_cbkeys_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_columns_with_field_names(df, genomincs_chars_df, exclude_cols=['eid']):\n",
    "    genomincs_chars_df = genomincs_chars_df.dropna(subset=['FieldID', 'Field']).copy()\n",
    "    genomincs_chars_df['FieldID'] = genomincs_chars_df['FieldID'].astype(int)\n",
    "    genomincs_chars_df['Field'] = genomincs_chars_df['Field'].str.strip()\n",
    "\n",
    "    fieldid_to_name = dict(zip(genomincs_chars_df['FieldID'], genomincs_chars_df['Field']))\n",
    "\n",
    "    rename_dict = {}\n",
    "    for col in df.columns:\n",
    "        if col in exclude_cols:\n",
    "            continue\n",
    "\n",
    "        match = re.match(r'f_(\\d+)_\\d+_\\d+', col)\n",
    "        if match:\n",
    "            fieldid = int(match.group(1))\n",
    "            new_name = fieldid_to_name.get(fieldid)\n",
    "            if new_name:\n",
    "                rename_dict[col] = new_name\n",
    "            else:\n",
    "                print(f\" FieldID {fieldid} encontrado en columna '{col}' no está en genomincs_chars_df\")\n",
    "        else:\n",
    "            print(f\" No coincide con patrón: {col}\")\n",
    "\n",
    "    return df.rename(columns=rename_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>Genotype measurement batch</th>\n",
       "      <th>Genetic sex</th>\n",
       "      <th>Heterozygosity</th>\n",
       "      <th>Heterozygosity, PCA corrected</th>\n",
       "      <th>Missingness</th>\n",
       "      <th>Genetic ethnic grouping</th>\n",
       "      <th>Genetic principal components</th>\n",
       "      <th>Genetic principal components</th>\n",
       "      <th>Genetic principal components</th>\n",
       "      <th>...</th>\n",
       "      <th>DNA concentration</th>\n",
       "      <th>Affymetrix quality control metric \"\"\"\"Cluster.CR\"\"\"\"</th>\n",
       "      <th>Affymetrix quality control metric \"\"\"\"dQC\"\"\"\"</th>\n",
       "      <th>Outliers for heterozygosity or missing rate</th>\n",
       "      <th>Use in phasing Chromosomes 1-22</th>\n",
       "      <th>Use in phasing Chromosome X</th>\n",
       "      <th>Use in phasing Chromosome XY</th>\n",
       "      <th>Genotype measurement plate</th>\n",
       "      <th>Genotype measurement well</th>\n",
       "      <th>HLA imputation values</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000012</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.186874</td>\n",
       "      <td>0.188240</td>\n",
       "      <td>0.008535</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-12.4522</td>\n",
       "      <td>2.97502</td>\n",
       "      <td>-1.526670</td>\n",
       "      <td>...</td>\n",
       "      <td>44.7400</td>\n",
       "      <td>99.059</td>\n",
       "      <td>0.97000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0010894A</td>\n",
       "      <td>G03</td>\n",
       "      <td>0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000021</td>\n",
       "      <td>-7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.190092</td>\n",
       "      <td>0.191694</td>\n",
       "      <td>0.001877</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-13.8955</td>\n",
       "      <td>6.01974</td>\n",
       "      <td>-2.011150</td>\n",
       "      <td>...</td>\n",
       "      <td>93.3443</td>\n",
       "      <td>99.477</td>\n",
       "      <td>0.95280</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0008357</td>\n",
       "      <td>B08</td>\n",
       "      <td>0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000035</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.190206</td>\n",
       "      <td>0.191802</td>\n",
       "      <td>0.001883</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-11.7626</td>\n",
       "      <td>4.23129</td>\n",
       "      <td>-3.091340</td>\n",
       "      <td>...</td>\n",
       "      <td>30.1373</td>\n",
       "      <td>99.595</td>\n",
       "      <td>0.96018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0008467</td>\n",
       "      <td>C11</td>\n",
       "      <td>1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000044</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.190249</td>\n",
       "      <td>0.191653</td>\n",
       "      <td>0.003119</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-11.2151</td>\n",
       "      <td>4.37638</td>\n",
       "      <td>0.766213</td>\n",
       "      <td>...</td>\n",
       "      <td>29.3400</td>\n",
       "      <td>99.580</td>\n",
       "      <td>0.98000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0011008A</td>\n",
       "      <td>A03</td>\n",
       "      <td>0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000053</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.190025</td>\n",
       "      <td>0.191251</td>\n",
       "      <td>0.002626</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-10.4144</td>\n",
       "      <td>2.19996</td>\n",
       "      <td>0.545220</td>\n",
       "      <td>...</td>\n",
       "      <td>22.1800</td>\n",
       "      <td>99.572</td>\n",
       "      <td>0.97000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SMP4_0013058A</td>\n",
       "      <td>H12</td>\n",
       "      <td>0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid  Genotype measurement batch  Genetic sex  Heterozygosity  \\\n",
       "0  1000012                        10.0          0.0        0.186874   \n",
       "1  1000021                        -7.0          1.0        0.190092   \n",
       "2  1000035                        -2.0          0.0        0.190206   \n",
       "3  1000044                        57.0          1.0        0.190249   \n",
       "4  1000053                        69.0          0.0        0.190025   \n",
       "\n",
       "   Heterozygosity, PCA corrected  Missingness  Genetic ethnic grouping  \\\n",
       "0                       0.188240     0.008535                      1.0   \n",
       "1                       0.191694     0.001877                      1.0   \n",
       "2                       0.191802     0.001883                      1.0   \n",
       "3                       0.191653     0.003119                      1.0   \n",
       "4                       0.191251     0.002626                      1.0   \n",
       "\n",
       "   Genetic principal components  Genetic principal components  \\\n",
       "0                      -12.4522                       2.97502   \n",
       "1                      -13.8955                       6.01974   \n",
       "2                      -11.7626                       4.23129   \n",
       "3                      -11.2151                       4.37638   \n",
       "4                      -10.4144                       2.19996   \n",
       "\n",
       "   Genetic principal components  ...  DNA concentration  \\\n",
       "0                     -1.526670  ...            44.7400   \n",
       "1                     -2.011150  ...            93.3443   \n",
       "2                     -3.091340  ...            30.1373   \n",
       "3                      0.766213  ...            29.3400   \n",
       "4                      0.545220  ...            22.1800   \n",
       "\n",
       "   Affymetrix quality control metric \"\"\"\"Cluster.CR\"\"\"\"  \\\n",
       "0                                             99.059      \n",
       "1                                             99.477      \n",
       "2                                             99.595      \n",
       "3                                             99.580      \n",
       "4                                             99.572      \n",
       "\n",
       "   Affymetrix quality control metric \"\"\"\"dQC\"\"\"\"  \\\n",
       "0                                        0.97000   \n",
       "1                                        0.95280   \n",
       "2                                        0.96018   \n",
       "3                                        0.98000   \n",
       "4                                        0.97000   \n",
       "\n",
       "   Outliers for heterozygosity or missing rate  \\\n",
       "0                                          NaN   \n",
       "1                                          NaN   \n",
       "2                                          NaN   \n",
       "3                                          NaN   \n",
       "4                                          NaN   \n",
       "\n",
       "   Use in phasing Chromosomes 1-22  Use in phasing Chromosome X  \\\n",
       "0                              1.0                          1.0   \n",
       "1                              1.0                          1.0   \n",
       "2                              1.0                          1.0   \n",
       "3                              1.0                          1.0   \n",
       "4                              1.0                          1.0   \n",
       "\n",
       "   Use in phasing Chromosome XY  Genotype measurement plate  \\\n",
       "0                           1.0               SMP4_0010894A   \n",
       "1                           1.0                SMP4_0008357   \n",
       "2                           1.0                SMP4_0008467   \n",
       "3                           1.0               SMP4_0011008A   \n",
       "4                           1.0               SMP4_0013058A   \n",
       "\n",
       "   Genotype measurement well  \\\n",
       "0                        G03   \n",
       "1                        B08   \n",
       "2                        C11   \n",
       "3                        A03   \n",
       "4                        H12   \n",
       "\n",
       "                               HLA imputation values  \n",
       "0  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...  \n",
       "1  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...  \n",
       "2  1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,...  \n",
       "3  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,...  \n",
       "4  0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gendre_df  = rename_columns_with_field_names(genomics_df, genomics_chars_df)\n",
    "\n",
    "gendre_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>Genetic sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000012</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000021</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000035</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000044</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000053</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid Genetic sex\n",
       "0  1000012      Female\n",
       "1  1000021        Male\n",
       "2  1000035      Female\n",
       "3  1000044        Male\n",
       "4  1000053      Female"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gendre_sex_df = gendre_df[['eid', 'Genetic sex']].copy()\n",
    "\n",
    "gendre_sex_df['Genetic sex'] = gendre_sex_df['Genetic sex'].map({0.0: 'Female', 1.0: 'Male'})\n",
    "\n",
    "gendre_sex_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(502180, 2)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gendre_sex_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gendre_sex_df.to_csv(os.path.join(path,\"genetic_sex.csv\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
